lib/core/src/Cardano/Pool/Metrics.hs

```diff
@@ -52,8 +73,105 @@ import GHC.Generics
+    deriving (Show, Eq)
+
+-- | @worker@ follows the chain and puts pool productions and stake
+-- distributions to a @DBLayer@, such that the data in the @DB@ always is
+-- consistent.
+--
+-- The pool productions and stake distrubtions in the DB can /never/ be from
+-- different forks.
+worker
+    :: NetworkLayer IO t Block
+    -> DBLayer IO
+    -> Trace IO Text
+    -> IO ()
+worker nl db tr = do
+    -- Read the k latest headers the DB knows about
+    let (_block0, bp) = staticBlockchainParameters nl
+    let k = fromIntegral . getQuantity . view #getEpochStability $ bp
+    knownHeaders <- readCursor db k
+
+    follow nl tr knownHeaders advance' rollback getHeader
+  where
+
+    advance'
+        :: NE.NonEmpty Block
+        -> BlockHeader
+        -> IO (FollowAction ErrMetricsWorkerErr)
+    advance' bs h = runExceptT (advance bs h) >>= return . \case
+        Left ErrMetricsWorkerErrTipTooVolatile
+            -> Retry
+        Left e
+            -> ExitWith e
+            -- TODO: Make all cases explicit here.
+        Right ()
+            -> Continue
+
+    advance
+        :: NE.NonEmpty Block
+        -> BlockHeader
+        -> ExceptT ErrMetricsWorkerErr IO ()
+    advance blocks nodeTip = do
+        -- Get a stake distribution and next batch of blocks such that we know
+        -- they are from the same chain.
+        --
+        -- Assumtion: We don't expect the node to switch from tip A to tip B,
+        -- and back to tip A.
+        --
+        -- If the tip is still the same after we have fetched the stake
+        -- distribution, we conclude that the stake-distrubtion is from the same
+        -- chain as the tip.
+        consistentDistr <- withinSameTip retryPolicy liftE getTip $ \nodeTip2 -> do
+            distr <- withExceptT (const ErrMetricsWorkerStakeDistributionUnavailible)
+                $ stakeDistribution nl
+            unless (nodeTip == nodeTip2) $
+                -- The tip has changed. We cannot guarantee that the blocks and
+                -- the stake-distribution are on the same chain.
+                throwE ErrMetricsWorkerErrTipTooVolatile
+            return (fst distr, Map.toList $ snd distr)
+
+        liftIO $ uncurry (putStakeDistribution db) consistentDistr
+
+        forM_ blocks $ \block ->
+            -- TODO: We might want to batch the insertion to the DB.
+            case producer block of
+                Just pool ->
+                    -- TODO: What do we do if the production already exists?!
+                    withExceptT (const ErrMetricsWorkerDataInconsistencyDuplicateBlock) $
+                        putPoolProduction db (header block) pool
+                Nothing ->
+                    -- We expect every block to have a producer. Ending up
+                    -- here would be /very/ unexpected.
+                    --
+                    -- TODO: How do we handle this?
+                    throwE ErrMetricsWorkerDataInconsistencyBlockWithoutProducer
```

Comment:
If we expect all blocks to have a producer here, then we should have the `Block` type to reflect that :+1:

Reply:
Would have to be something like `Either Genesis Block` in that case. Maybe that's good.

Reply:
Right, of course `Either Genesis Block` doesn't fix the problem. It just re-packages the `Nothing` case as a `Genesis`.

We would have to parameterise `NetworkLayer` by both block0 and `block`… 🤔

Reply:
Specialised `block0` to `W.Block Tx` as discuseed.